File: back-end/src/main/scala/com/example/lumina/Domain/Span.scala
package com.example.lumina.Domain

import io.circe.{Json, Encoder, Decoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
import com.example.lumina.types.{SpanKind, SpanStatus}
import com.example.lumina.types.given

import java.time.OffsetDateTime
import java.util.UUID

case class Span(
    id: UUID,
    traceId: UUID,
    parentSpanId: Option[UUID],
    name: String,
    kind: SpanKind,
    status: SpanStatus,
    error: Option[String],
    startedAt: OffsetDateTime,
    endedAt: Option[OffsetDateTime],
    durationMs: Option[Int],
    input: Json,
    output: Json,
    attributes: Json
)

object Span {
  given Encoder[Span] = deriveEncoder[Span]
  given Decoder[Span] = deriveDecoder[Span]
}

File: sdk/typescript/src/API/span/span.ts
import { UUID } from "node:crypto"
import { SpanKind, SpanStatus } from "../types"

export interface Span {
    id: UUID
    traceId: UUID
    parentSpanId: UUID | null
    name: string
    kind: SpanKind
    status: SpanStatus
    error: string | null
    startedAt: Date
    endedAt: Date | null
    durationMs: number | null
    input: Record<string, unknown>
    output: Record<string, unknown>
    attributes: Record<string, unknown>
}

export interface StartSpanBody<T, K> {
    name: string
    kind: SpanKind
    input: T
    attributes?: Record<string, unknown>
    callback: (input: T) => Promise<K>
}